File: Graph Algorithms/building_teams.cpp
#include<bits/stdc++.h>
#include <sys/resource.h>
#define INF 1000000000
#define endl '\n'
const double PI = 3.141592653589793238460;
const int mxN = 2e5;
using namespace std;

class Vertex{
    public:
    int data;
    int visited;
    int group;
    vector<int> connections;
};

int dfs(unordered_map<int, Vertex>& graph, int n){
    int count = 0;
    stack<Vertex> s;
    for(int i = 1; i <= n; i++){
        if(graph[i].visited) continue;
        count++;
        s.push(graph[i]);
        graph[i].group = count;
        while(!s.empty()){
            Vertex v = s.top();
            s.pop();
            for(auto i: v.connections){
                if(graph[i].visited) continue;
                graph[i].visited = 1;
                graph[i].group = count; 
                s.push(graph[i]);
            }
        }
        
    }
    return count;
}
int main(){
    rlimit cpu_time{.rlim_cur = 1, .rlim_max = 5}; 
    setrlimit(RLIMIT_CPU, &cpu_time);
    ios_base::sync_with_stdio(0);
    int n, m,a,b;
    cin >> n >> m;
    unordered_map<int , Vertex> graph;
    for(int i = 0; i < n; i++){
        graph[i+1] = Vertex();
        graph[i+1].data = i+1;
        graph[i+1].visited = 0;
        graph[i+1].group = 0;
    }
    for(int i = 0 ; i < m; i++){
        cin >> a >> b;
        graph[a].connections.push_back(b);
        graph[b].connections.push_back(a);
    }
    int count = dfs(graph,n);
    if(count != 2){
        cout << "IMPOSSIBLE" << endl;
    }else{
        for(int i = 1; i <= n; i++){
            cout << graph[i].group << " ";
        }
        cout << endl;
    }
}

File: Graph Algorithms/building_teams.py
class Vertex:
    def __init__(self,data):
        self.data = data 
        self.visited = False
        self.group = 1 
        self.connections = []

# fails 4 testcases
def solve():
    n, m = map(int,input().split())
    graph = dict()
    for i in range(1,n+1):
        graph[i] = Vertex(i)
    for _ in range(m):
        a,b = map(int,input().split())
        graph[a].connections.append(b)
        graph[b].connections.append(a)
    flag = False 
    for i in range(1,n+1):
        for j in graph[i].connections:
            if j <= i and graph[i].group == graph[j].group:
                flag = True
                break 
            if graph[i].group == graph[j].group:
                graph[j].group = 0 if graph[i].group else 1

    if flag:
        print("IMPOSSIBLE")
        return 

    for i in range(1,n+1):
        print(graph[i].group+1,end = " ")
    print()

if __name__ == '__main__':
    solve() 
